db/similarity.py

```diff
@@ -0,0 +1,239 @@
+
+        while count_similarity() < lowlevel_count:
+            batch_query = text("""
+                SELECT id
+                  FROM lowlevel
+              ORDER BY id
+                 LIMIT :batch_size
+                OFFSET :offset
+            """)
+            result = connection.execute(batch_query, {"batch_size": batch_size, "offset": offset})
+            if not result.rowcount:
+                print("Metric {} added for all recordings".format(metric.name))
+                break
+
+            with connection.begin():
+                for row in result.fetchall():
+                    vectors_info = []
+                    for metric, data in get_metrics_data(row["id"], metrics):
+                        try:
+                            vector = metric.transform(data)
+                            isnan = False
+                        except ValueError:
+                            vector = [None] * metric.length()
+                            isnan = True
+                        vectors_info.append((metric.name, vector, isnan))
+                    insert_similarity(connection, row["id"], vectors_info)
+
+            offset = count_similarity()
+            print("Processed {} / {} ({:.3f}%)".format(offset,
+                                                       lowlevel_count,
+                                                       float(offset) / lowlevel_count * 100))
+
+
+def get_metrics_data(id, metrics):
+    ret = []
+    for metric in metrics:
+        data = metric.get_data(id)
+        ret.append((metric, data))
+    return ret
+
+
+def insert_similarity(connection, id, vectors_info):
+    # vectors_info = [(metric_name, vector, isnan)]
+    values = []
+    for metric, vector, isnan in vectors_info:
+        value = ('ARRAY' + ('[' + ', '.join(["'NaN'::double precision"] *
+                 len(vector)) + ']' if isnan else str(list(vector))))
+        values.append(value)
+
+    values_string = ', '.join(values)
+
+    query = text("""
+        INSERT INTO similarity (
+                    id,
+                    mfccs,
+                    mfccsw,
+                    gfccs,
+                    gfccsw,
+                    key,
+                    bpm,
+                    onsetrate,
+                    moods,
+                    instruments,
+                    dortmund,
+                    rosamerica,
+                    tzanetakis)
+             VALUES (
+                    :id,
+                    %(values)s)
+        ON CONFLICT (id)
+         DO NOTHING
+    """ % {"values": values_string})
+    connection.execute(query, {'id': id})
+
+
+def count_similarity():
+    # Get total number of submissions in similarity table
+    with db.engine.connect() as connection:
+        query = text("""
+            SELECT COUNT(*)
+              FROM similarity
+        """)
+        result = connection.execute(query)
+        return result.fetchone()[0]
+
+
+def submit_similarity_by_id(id):
```

Comment:
@alastair This function and the one below are meant to submit only a single row, this will be necessary when we need to update the table with new submissions.